### PR TITLE
Remove backwards-compatibility shims and deprecated aliases

### PR DESCRIPTION
**Pull request by imreddyTeja**
The DataLayouts, Geometry and Operators rewrites of the last few releases each left a compatibility layer behind, to be deleted "once downstream consumers have migrated". ClimaAtmos and ClimaLand have now migrated, so this deletes them.

Every removed name had a direct replacement; nothing else about the API changes.

**Deletes** `src/deprecated.jl`, `src/{DataLayouts,Geometry,Operators}/deprecated.jl`, `src/recursive_apply.jl` and `test/deprecations.jl`.

## Removed, and what replaces it

| Removed | Replacement |
|---|---|
| `Geometry.AxisTensor{T, N, B, S}` | `Tensor{N, T, B, S}` (T and N swapped) |
| `Geometry.AxisVector{T, A, S}` | `Tensor{1, T, Tuple{A}, S}` |
| `Geometry.Axis2Tensor{T, B, S}` | `Tensor{2, T, B, S}` |
| `Geometry.AxisTensor(bases, comps)` | `Tensor(comps, bases)` |
| `Geometry.components(x)` for tensors | `parent(x)` |
| `{Covariant,Contravariant,Local,Cartesian}Axis{I}` | the concrete aliases that name their indices, e.g. `Contravariant12Axis` |
| `DataLayouts.AbstractData{T}` | `DataLayout{T}` |
| `DataLayouts.IJFH{T, Nij}` / `IJHF{T, Nij}` | `VIJFH{T, 1, Nij, Nij, nothing}` / `VIJHF{…}` |
| 5-D `CartesianIndex` layout indexing | `CartesianIndex(v, i, j, h)` |
| `Operators.Weak{Divergence,Gradient,Curl}` | `{Divergence,Gradient,Curl}{WeakForm}` |
| `Operators.{Left,Right}Biased{C2F,F2C}` | `{Bottom,Top}Biased{C2F,F2C}` |
| `Operators.{First,Third}OrderOneSided` | `Extrapolate{0}` / `Extrapolate{1}` |
| `Fields.ColumnField` | `FiniteDifferenceField` |
| no-index `Fields.level(field)` | `level(field, v)` |
| `PointCloud{Grid,Space}`, `ExtrudedPointCloudGrid`, `PointColumnEnsemble{Grid,Space}` | the `Multi*` names |
| `MatrixFields.⋅` | `*` (matrix multiplication since v0.14.27) |
| `IntervalTopology(mesh)` and the three sibling mesh-only constructors | the `(device, mesh)` forms |
| the `RecursiveApply` module | `Utilities.add_auto_broadcasters` / `drop_auto_broadcasters` |

**Exports dropped:** `AbstractData`/`IJFH`/`IJHF` from `DataLayouts`, `⋅` from `MatrixFields`, and `PointColumnEnsembleGrid`/`PointColumnEnsembleSpace` from `CommonGrids`/`CommonSpaces`.

### CODE REVIEW

File: src/Operators/common.jl [hunk 1]
```diff
@@ -247,3 +247,28 @@ function unstrip_space(bc::Broadcast.Broadcasted, parent_space)
...
+
+# Read the value of `field` at one node of a slab. Used by the `Remapping`
+# module (and its CUDA extension) to interpolate a field onto arbitrary points.
+# A 1-dimensional index is a column of a 1D space, so its second index is 1.
+Base.@propagate_inbounds function get_node(
+    parent_space,
+    field::Fields.Field,
+    ij::CartesianIndex{N},
+    slabidx,
+) where {N}
```
Comment (dennisYatunin) [inline src/Operators/common.jl:259]: It would be good to update the `Remapping` module to the new `field[v, i, j, h]` syntax, instead of keeping this outdated function. Or at least simplify this and move it into the `Remapping` module, since it's more of a remapping helper function than an `Operators` API function.

File: test/DataLayouts/unit_fill_and_copyto.jl [hunk 2]
```diff
@@ -59,7 +64,7 @@ function test_single_F!(data)
     @test all(to_data(parent(data)) .== to_data(parent(rand_data)))
 
     Base.copyto!(data, Base.Broadcast.broadcasted(+, rand_data, 0x1))
-    @test all(to_data(parent(data)) .== to_data(parent(rand_data)) .⊞ 0x1)
+    @test all(to_data(parent(data)) .== nested_add.(to_data(parent(rand_data)), 0x1))
```
Comment (dennisYatunin) [inline test/DataLayouts/unit_fill_and_copyto.jl:67]: This would be simpler if `nested_add` were replaced with `+`, which you could do by adding `AutoBroadcaster` wrappers in `to_data`:
```julia
to_data(array) = DataLayouts.bitcast_struct.(eltype(Base.broadcastable(data)), array)
```
That way, you could drop the `nested_add` definition and the new imports from `ClimaCore.Utilities`.

File: src/MatrixFields/operator_matrices.jl [hunk 3]
```diff
@@ -81,6 +81,12 @@ operator_input_space(
     space::Spaces.MultiColumnFiniteDifferenceSpace,
 ) = space
 
+# The zero of a row type whose entries may themselves be composite (e.g. a
+# `NamedTuple`-valued band matrix row), which `Base.zero` cannot build on its
+# own. Wrapping the type in an `AutoBroadcaster` maps `zero` over the entries.
+nested_zero(::Type{T}) where {T} =
+    drop_auto_broadcasters(zero(add_auto_broadcasters(T)))
```
Comment (dennisYatunin) [inline src/MatrixFields/operator_matrices.jl:88]: Instead of defining `nested_zero` here, you could extend `Base.broadcastable` to operator boundary conditions, and call `Base.broadcastable(op)` when the broadcast expression is constructed. That should allow you to use the standard `zero` function below.

File: test/MatrixFields/unit_operator_matrices.jl [hunk 4]
```diff
@@ -1,6 +1,11 @@
 import LinearAlgebra: I
 
-import ClimaCore.RecursiveApply: rzero
+import ClimaCore.Utilities: add_auto_broadcasters, drop_auto_broadcasters
+
+# `zero` of a type whose entries may themselves be composite, which `Base.zero`
+# cannot build on its own.
+zero_of(::Type{T}) where {T} =
+    drop_auto_broadcasters(zero(add_auto_broadcasters(T)))
```
Comment (dennisYatunin) [inline test/MatrixFields/unit_operator_matrices.jl:8]: This might also become unnecessary if you extend `Base.broadcastable` to handle operator boundary conditions.

**Review by dennisYatunin — APPROVED** [review]

**Review by dennisYatunin — APPROVED** [review]
